src/librepgp/stream-parse.c

```diff
@@ -568,114 +503,183 @@ encrypted_src_close(pgp_source_t *src)
 {
-    ssize_t len;
+    pgp_source_encrypted_param_t *param = src->param;
+    pgp_crypt_t                   crypt;
+    uint8_t                       enchdr[PGP_MAX_BLOCK_SIZE + 2];
+    unsigned                      blsize;
 
-    len = stream_read_pkt_len(src);
-    if (len < 0) {
-        return RNP_ERROR_READ;
+    if (!(blsize = pgp_block_size(alg))) {
+        return false;
     }
 
-    (void) fprintf(stderr, "skipping public key encrypted session key\n");
-    src_skip(src, len);
-    return RNP_SUCCESS;
-}
-
-static rnp_result_t
-stream_parse_sk_sesskey(pgp_source_t *src, pgp_sk_sesskey_t *skey)
-{
-    uint8_t buf[4];
-    ssize_t len;
-    ssize_t read;
-
-    // read packet length
-    len = stream_read_pkt_len(src);
-    if (len < 0) {
-        return RNP_ERROR_READ;
-    } else if (len < 4) {
-        return RNP_ERROR_BAD_FORMAT;
+    /* reading encrypted header to check the password validity */
+    if (src_peek(param->pkt.readsrc, enchdr, blsize + 2) < blsize + 2) {
+        (void) fprintf(stderr, "%s: failed to read encrypted header\n", __func__);
+        return false;
     }
 
-    // version + symalg + s2k type + hash alg
-    if ((read = src_read(src, buf, 4)) < 4) {
-        return RNP_ERROR_READ;
+    /* having symmetric key in keybuf let's decrypt blocksize + 2 bytes and check them */
+    if (!pgp_cipher_start(&crypt, alg, key, NULL)) {
+        (void) fprintf(stderr, "%s: failed to start cipher\n", __func__);
+        return false;
     }
 
-    // version
-    skey->version = buf[0];
-    if (skey->version != 4) {
-        (void) fprintf(stderr, "stream_parse_sk_sesskey: wrong packet version\n");
-        return RNP_ERROR_BAD_FORMAT;
-    }
+    pgp_cipher_cfb_decrypt(&crypt, enchdr, enchdr, blsize + 2);
+    if ((enchdr[blsize] == enchdr[blsize - 2]) && (enchdr[blsize + 1] == enchdr[blsize - 1])) {
+        src_skip(param->pkt.readsrc, blsize + 2);
+        param->decrypt = crypt;
+        /* init mdc if it is here */
+        /* RFC 4880, 5.13: Unlike the Symmetrically Encrypted Data Packet, no special CFB
+         * resynchronization is done after encrypting this prefix data. */
+        if (!param->has_mdc) {
+            pgp_cipher_cfb_resync(&param->decrypt);
+        } else {
+            if (!pgp_hash_create(&param->mdc, PGP_HASH_SHA1)) {
+                pgp_cipher_finish(&crypt);
+                (void) fprintf(stderr, "%s: cannot create sha1 hash\n", __func__);
+                return false;
+            }
 
-    // symmetric algorithm
-    skey->alg = buf[1];
+            pgp_hash_add(&param->mdc, enchdr, blsize + 2);
+        }
 
-    // s2k
-    skey->s2k.specifier = buf[2];
-    skey->s2k.hash_alg = buf[3];
-    len -= 4;
+        return true;
+    } else {
+        return false;
+    }
+}
 
-    switch (skey->s2k.specifier) {
-    case PGP_S2KS_SIMPLE:
+static bool
+encrypted_try_key(pgp_source_t *src, pgp_pk_sesskey_pkt_t *sesskey, pgp_key_t *key)
```

Comment:
This function is partially copy paste of ```pgp_create_pk_sesskey()```. Can you reuse the code which already exist?

Reply:
Yeah, I borrowed code from old places. The reason I don't reuse old functions is that they use other input/output model and don't follow the 'dump entities' logic, i.e. packet struct is just a parsed OpenPGP entity without any additional data like decrypted keys, flags, and so on. Once transition to the streams is finished all old code and old functions will be removed.

Reply:
ok. I see, that's just for transition time

Reply:
Agree, added #490